File: .archive/FoodLocations.cpp
//
//  FoodLocations.cpp
//  stunning-potato
//
//  Created by Daniel Liu on 20-5-29.
//  Copyright © 2020 Daniel Liu. All rights reserved.
//

#include <stdio.h>
#include <algorithm>
#include "FoodLocations.h"

using namespace std;

FoodLocations::FoodLocations() : places{}, namelookup{}, ratinglookup{41, vector<int>()}, pricelookup{4, vector<int>()} {}

void FoodLocations::addPlace(const FoodPlace &fp) {
    places.push_back(fp);
    namelookup[fp.name] = (int) places.size() - 1;
    for (auto type : fp.types) {
        typelookup[type].push_back((int) places.size() - 1);
    }
}

int FoodLocations::searchPlace(const string &name) {
    return (namelookup.find(name) != namelookup.end()) ? namelookup[name] : -1;
}

template <typename T>
void append(vector<T> &original, const vector<T>& add) {
    for (int x : add) {
        original.push_back(x);
    }
}

/*
vector<int> FoodLocations::searchTypes(const vector<string> &types) {
    vector<int> results;
    for (const string &type : types) {
        append(results, typelookup[type]);
    }
    return results;
}
*/

vector<int> FoodLocations::ratingGreaterThan(int num) {
    vector<int> results;
    int look = 40;
    while (look + 10 >= num) {
        append(results, ratinglookup[look]);
        --look;
    }
    return results;
}

vector<int> FoodLocations::priceLTEqualTo(int num) {
    vector<int> pl;
    for (int i = num - 1; num >= 0; --num) {
        append(pl, pricelookup[i]);
    }
    sort(pl.begin(), pl.end(), [this](int a, int b) {
        return FoodPlaceComp()(places[b], places[a]);
    });
    return pl;
}

File: StunningPotato/public/utils/foodLocations.js
class FoodLocations {
    constructor(socket) {
        this.socket = socket;
        this.places = [];
        this.nameLookup = {};
        this.typeLookup = {};
        this.ratingLookup = [];
        for (let i = 0; i < 41; ++i) this.ratingLookup.push([]);
        this.priceLookup = [];
        for (let i = 0; i < 4; ++i) this.priceLookup.push([]);
    }

    addPlace(place) {
        this.places.push(place);
        this.nameLookup[place.name] = this.places.length - 1;
        //for (let i = 0; i < place.types.length; ++i) {
            //this.typeLookup[place.types[i]].push(this.places.length - 1);
        //}
        if (place.price == null) {
            place.price = 3;
        }
        if (place.rating == null) {
            place.rating = 1;
        }
        this.priceLookup[Math.floor(place.price) - 1].push(place);
        //this.ratingLookup[Math.floor(place.rating * 10) - 10].push(place);
    }

    searchPlace(name) {
        return (name in this.nameLookup ? this.nameLookup[name] : -1);
    }

    ratingGreaterThan(num) {
        var results = [];
        let look = 40;
        while (look + 10 >= num) {
            for (let i = 0; i < this.ratingLookup[look].length; ++i) {
                results.push(this.ratingLookup[look]);
            }
            --look;
        }
        return results;
    }

    priceLTEqualTo(num) {
        var p = [];
        for (let i = num - 1; i >= 0; --i) {
            for (let j = 0; j < this.priceLookup[i].length; ++j) {
                p.push(this.priceLookup[i][j]);
            }
        }
        p.sort(function(a, b) { return
        (b.rating < a.rating)});
        return p;
    }

}

module.exports = {
    FoodLocations: FoodLocations
}